src/gnark_backend_wrapper/plonk/mod.rs

```diff
@@ -0,0 +1,252 @@
+use crate::acvm;
+use std::collections::BTreeMap;
+use std::ffi::{CStr, CString};
+use std::num::TryFromIntError;
+use std::os::raw::{c_char, c_uchar};
+
+use crate::gnark_backend_wrapper::c_go_structures::{GoString, KeyPair};
+use crate::gnark_backend_wrapper::errors::GnarkBackendError;
+use crate::Gnark;
+
+// Arkworks's types are generic for `Field` but Noir's types are concrete and
+// its value depends on the feature flag.
+cfg_if::cfg_if! {
+    if #[cfg(feature = "bn254")] {
+        pub use ark_bn254::{Bn254 as Curve, Fr};
+
+        // Converts a FieldElement to a Fr
+        // noir_field uses arkworks for bn254
+        pub fn from_felt(felt: acvm::FieldElement) -> Fr {
+            felt.into_repr()
+        }
+    } else if #[cfg(feature = "bls12_381")] {
+        pub use ark_bls12_381::{Bls12_381 as Curve, Fr};
+
+        // Converts a FieldElement to a Fr
+        // noir_field uses arkworks for bls12_381
+        pub fn from_felt(felt: FieldElement) -> Fr {
+            felt.into_repr()
+        }
+    } else {
+        compile_error!("please specify a field to compile with");
+    }
+}
+
+extern "C" {
+    fn PlonkVerifyWithMeta(circuit: GoString, values: GoString, proof: GoString) -> c_uchar;
+    fn PlonkProveWithMeta(circuit: GoString, values: GoString) -> *const c_char;
+    fn PlonkVerifyWithVK(rawr1cs: GoString, proof: GoString, verifying_key: GoString) -> c_uchar;
+    fn PlonkProveWithPK(
+        circuit: GoString,
+        values: GoString,
+        proving_key: GoString,
+    ) -> *const c_char;
+    fn PlonkPreprocess(circuit: GoString, values: GoString) -> KeyPair;
+}
+
+pub fn prove_with_meta(
+    circuit: acvm::Circuit,
+    values: Vec<acvm::FieldElement>,
+) -> Result<Vec<u8>, GnarkBackendError> {
+    // Serialize to json and then convert to GoString
+    let acir_json = serde_json::to_string(&circuit)
+        .map_err(|e| GnarkBackendError::SerializeCircuitError(e.to_string()))?;
+    let acir_c_str = CString::new(acir_json)
+        .map_err(|e| GnarkBackendError::SerializeCircuitError(e.to_string()))?;
+    let acir_go_string = GoString::try_from(&acir_c_str)?;
+
+    let values_json = serde_json::to_string(&circuit)
```

Comment:
Here and in the other methods that receive `values` as a parameter we are converting `circuit` to JSON instead of `values`. Anyways, I didn't make a code suggestion because we cannot serialize `values` to JSON directly like this. The values should be an encoded slice of bytes like they are in `RawR1CS`. For this we'll need to "cast" `acvm::FieldElement` to a `Fr`, for this you could use `from_felt`, and then we should use the `serialize_felts` function.